Catalog zones

wtoorop opened the pull request:
This version does the task from within the tranfer daemon (xfrd) which makes everything easier and more straight forward.

wtoorop committed:
Configure options for catalog zones

wtoorop committed:
Set ident (and proctitle) earlier for debugging

Also change ident to "load" when reloading and the backup main process
to "old-main". When reload succeeds, "load" will become "main" and
"old-main" will quit. Otherwise "load" will be exited (because of
failure) and "old_main" will continue as "main".

wtoorop committed:
dname_to_string_buf()

A version of dname_to_string() that writes to a buffer instead of
returning a static buffer. This is useful when you need to print two
dnames in a log message.

wtoorop committed:
Basic catalog consumer zone processing

TODO: Group property

wtoorop committed:
The group property

wtoorop committed:
Be explicit about the catalog role, also in the config

+ marshalling of the catalog zone pattern options

wtoorop committed:
Fix for copy/paste bug found by clang

k0ekk0ek commented on nsd.conf.5.in:
```diff
@@ -990,10 +990,22 @@ Feed updated zone to verifier over standard input. Default is
+.TP
+.B catalog:\fR <consumer or producer>
+If set to consumer, catalog zone processing is enabled for this catalog.
+Member zones of the catalog will use the patter specified by the group
```
missing n in patter(n)

wtoorop replied:
Fixed in 1fce7ed1bdf1cccba8c35e27fb046e786651aa1a

wtoorop commented on nsd.conf.5.in:
```diff
@@ -990,10 +990,22 @@ Feed updated zone to verifier over standard input. Default is
+.TP
+.B catalog:\fR <consumer or producer>
+If set to consumer, catalog zone processing is enabled for this catalog.
+Member zones of the catalog will use the patter specified by the group
```
```suggestion
Member zones of the catalog will use the pattern specified by the group
```

wtoorop committed:
missing n in pattern in nsd.conf.5.in

pettai commented:
Current version works fine for us 👍  (just replaced one of our test-installations of a previous catz-branch build)

wtoorop committed:
Merge branch 'master' into features/xfrd-catz

wtoorop committed:
Merge branch 'features/xfrd-catz' of github.com:NLnetLabs/nsd into features/xfrd-catz

pettai commented:
First issue encountered:
`Nov 27 20:10:01 sunic nsd[2498035]: invalid catalog consumer zone 'test.catalog': 'version.test.catalog TXT RRset not found`

```
zone:	test.catalog
	catalog: consumer
	catalog-invalid: 'version.test.catalog TXT RRset not found
	state: ok
	served-serial: "1698045664 since 2023-11-27T19:31:01"
	commit-serial: none
	wait: "84395 sec between attempts"
```

A manual `nsd-control transfer test.catalog` fixed the issue.

pettai commented:
Next issue I see then I restart `nsd` with `systemctl restart nsd`. There are some defunct nsd processes around, but nothing insightful in the logs. (nsd with the catzones seems to work fine...)

```
 22:08 root@sunic: /etc/nsd/nsd.conf.d # ps -ef | grep nsd
nsd      2850692       1  0 22:07 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2850694 2850692 98 22:07 ?        00:01:12 /usr/sbin/nsd -d -P
nsd      2853485 2850694  3 22:08 ?        00:00:00 [nsd: server 2] <defunct>
nsd      2853486 2850694  3 22:08 ?        00:00:00 [nsd: server 3] <defunct>
nsd      2853487 2850694  3 22:08 ?        00:00:00 [nsd: server 4] <defunct>
nsd      2853488 2850694  3 22:08 ?        00:00:00 [nsd: server 5] <defunct>
nsd      2853490 2850694  3 22:08 ?        00:00:00 [nsd: server 7] <defunct>
nsd      2853491 2850694  3 22:08 ?        00:00:00 [nsd: server 8] <defunct>
nsd      2853492 2850694  3 22:08 ?        00:00:00 [nsd: server 9] <defunct>
nsd      2853498 2850694  3 22:08 ?        00:00:00 [nsd: server 15] <defunct>
nsd      2853507 2850694  4 22:08 ?        00:00:00 [nsd: server 24] <defunct>
nsd      2853508 2850694  4 22:08 ?        00:00:00 [nsd: server 25] <defunct>
nsd      2853510 2850694  4 22:08 ?        00:00:00 [nsd: server 27] <defunct>
nsd      2853511 2850694  4 22:08 ?        00:00:00 [nsd: server 28] <defunct>
nsd      2853512 2850694  4 22:08 ?        00:00:00 [nsd: server 29] <defunct>
nsd      2853513 2850694  4 22:08 ?        00:00:00 [nsd: server 30] <defunct>
nsd      2853516 2850694  4 22:08 ?        00:00:00 [nsd: old-main] <defunct>
nsd      2853519 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853520 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853521 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853522 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853523 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853524 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853525 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853526 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853527 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853528 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853529 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853530 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853531 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853532 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853533 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853534 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853535 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853536 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853537 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853538 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853539 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853540 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853541 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853542 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853543 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853544 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853545 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853546 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853547 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853548 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853549 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853550 2850694  0 22:08 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853551 2850694  0 22:09 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853552 2850694  0 22:09 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853553 2850694  0 22:09 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853554 2850694  0 22:09 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853555 2850694  0 22:09 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853556 2850694  0 22:09 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853557 2850694  0 22:09 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853558 2850694  0 22:09 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853559 2850694  0 22:09 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853560 2850694  0 22:09 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853561 2850694  0 22:09 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853562 2850694  0 22:09 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853563 2850694  0 22:09 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853564 2850694  0 22:09 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853565 2850694  0 22:09 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853566 2850694  0 22:09 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853567 2850694  0 22:09 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853568 2850694  0 22:09 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853569 2850694  0 22:09 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853570 2850694  0 22:09 ?        00:00:00 /usr/sbin/nsd -d -P
nsd      2853571 2850694  0 22:09 ?        00:00:00 /usr/sbin/nsd -d -P
root     2853573 2837108  0 22:09 pts/0    00:00:00 grep --color=auto nsd
```

pettai commented:
> Next issue I see then I restart `nsd` with `systemctl restart nsd`. There are some defunct nsd processes around, but nothing insightful in the logs. (nsd with the catzones seems to work fine...)

This seems to be related to configuration processing (like the previous catz work)

```
pattern:
	name: "test.catalog"
	zonefile: "/var/lib/nsd/catzones/%s"

zone:
	name: "test.catalog."   <--- by adding a tailing dot (.) the <defunct> processes goes away
	catalog: consumer
	catalog-member-pattern: "test.catalog"
```

wtoorop committed:
struct dname for member_id

instead of struct domain

wtoorop committed:
Some code consistency

wtoorop committed:
Merge branch 'master' into features/xfrd-catz

wtoorop committed:
One forgotten code consistency

wtoorop committed:
Basic catalog zone producer

TODO:
- Check all cases where zones are added, deleted or changed for status
  change of: producer zones and producer member zones.
- A zone could be a consumer member zone, *and* a producer zone or a
  producer member.

wtoorop committed:
Merge branch 'master' into features/xfrd-catz-producer

wtoorop committed:
Merge branch 'features/xfrd-catz-producer' into features/xfrd-catz

wtoorop committed:
Show serial and # members for catalog zones in nsd-control zonestatus,

and process more than one transfer file.
TODO change processing after successful SOAINFO
	... before xfrd_unlink_xfrfile().

wtoorop committed:
Apply xfr to catalog consumer when confirmed with SOAINFO, and...

... check zonefile when asked to check.

wtoorop committed:
Only exit apply_ixfr_for_zone() when called from server process

The function can still exit via apply_ixfr() -> addRR() or deleteRR()
when out of memory, but in that case all bets are off anyway, right?

wtoorop committed:
Mitigate older clang analyzer false positive

wtoorop committed:
Mitigate older clang false positive with assert

wtoorop committed:
Move catalog zone processing to xfrd-catalog-zones.[ch]

wtoorop committed:
updated dependencies for catalog zone processing

wtoorop committed:
More config checks for catalog zones

+ some fixes

wtoorop committed:
Put xfrd catalog zone processing in separate files

wtoorop committed:
xfrd_producer_member tracks adding and deletion...

... of producer member zones, for constructing the artificial transfer file

wtoorop committed:
Disable some oververbose debugging

wtoorop committed:
consumer member zones in rbtree...

instead of double linked list

wtoorop committed:
Some fixes for adding and deleting member zones

wtoorop committed:
Fix deletion of remaining zones

wtoorop committed:
fix apply_ixfr_for_zone() returns 0 on error

wtoorop committed:
Apply multiple potential xfrs to a catalog consumer zone

wtoorop committed:
Fix adding and deleting the same member zones with different member_ids

wtoorop committed:
Explain catalog consumer processing modes in comment

wtoorop committed:
Force transfer consumer zones that become valid

wtoorop committed:
producer member zone without producer is just a zone

wtoorop committed:
Happy path catalog zone testing

wtoorop committed:
please clang

wtoorop committed:
Some things from travis build

wtoorop committed:
Updated nsd.conf man page and example for catalog zones

wtoorop committed:
Some config error checking for catalog producer zones

and catalog producer member zones

wtoorop committed:
Final configuring catalog zones tests

Everything is ready for review now

wtoorop committed:
test catalog consumer member zone cannot be deleted

wcawijngaards commented on nsd-checkconf.c:
```diff
@@ -585,6 +585,19 @@ static void print_zone_content_elems(pattern_options_type* pat)
 		printf("\tverifier-timeout: %d\n", pat->verifier_timeout);
 	}
+
+	switch(pat->catalog_role) {
```
Should this check if the setting is at default, like for max-refresh-time? Suggestion for that:
```suggestion
	if(!pat->catalog_role_is_default)
	    switch(pat->catalog_role) {
```

wcawijngaards commented on nsd.conf.5.in:
```diff
@@ -987,10 +987,52 @@ Feed updated zone to verifier over standard input. Default is
+by the \fBcatalog-member-pattern\fR option is used. Group properties are valid
+if there is only a single value matching the name of a for member zones valid
+pattern.
+
```
The empty line means a break is desired, so a suggested change for it.
```suggestion
.IP
```

wcawijngaards commented on nsd.conf.5.in:
```diff
@@ -987,10 +987,52 @@ Feed updated zone to verifier over standard input. Default is
+\fBrequest\-xfr\fR option. Catalog producer zones will \fInot\fR read content
+from zone files, but will reconstruct the zone on startup from the member zone
+entries in @zonelistfile@, specified with the \fBzonelistfile\fR option.
+
```
```suggestion
.IP
```

wcawijngaards commented on nsd.conf.5.in:
```diff
@@ -987,10 +987,52 @@ Feed updated zone to verifier over standard input. Default is
+the catalog zone is invalid, the reason for it to be invalid is shown.
+\fInsd\-control zonestatus\fR will also show the entry of a catalog member zone
+in the catalog (consumer or producer) zone as \fBcatalog-member-id:\fR.
+
```
For paragraph.
```suggestion
.IP
```

wcawijngaards commented on options.c:
```diff
@@ -465,8 +548,42 @@ parse_zone_list_file(struct nsd_options* opt)
+			*mem_idnm++ = 0;
+			patnm = strrchr(nm, ' ');
+			if(!mem_idnm) {
+				*--mem_idnm = 0;
```
Since mem_idnm is already checked, and it just fetched strrchr on patnm, I guess this one needs to be checked here. And the line on failure is perhaps to restore the space for the error printout to print the entire line.
```suggestion
			if(!patnm) {
				*--mem_idnm = ' ';
```

wtoorop replied:
Yes that was the intention! Thanks for catching this!

wcawijngaards commented on options.h:
```diff
@@ -488,12 +518,16 @@ void tls_auth_options_insert(struct nsd_options* opt, struct tls_auth_options* a
-struct zone_options* zone_list_add(struct nsd_options* opt, const char* zname,
-	const char* pname);
+static inline struct zone_options* zone_list_add(struct nsd_options* opt,
+		const char* zname, const char* pname)
```
It seems based on the other indentation, that one tab is intended here, for the function declaration line.
```suggestion
	const char* zname, const char* pname)
```

wcawijngaards commented on remote.c:
```diff
@@ -1037,6 +1040,62 @@ print_zonestatus(RES* ssl, xfrd_state_type* xfrd, struct zone_options* zo)
+			return 0;
+		if(producer_zone) {
+			if(!ssl_printf(ssl, " (serial: %u, # members: %zu)\n",
+					producer_zone->serial,
```
In other places, there is a cast to print serial numbers as unsigned, possibly due to platforms or compilers complaining about the integer type and size, the suggested edit adds it.
```suggestion
					(unsigned)producer_zone->serial,
```

wtoorop committed:
Merge branch 'master' into features/xfrd-catz

wcawijngaards commented on tpkg/catzones.tdir/catzones.test:
```diff
@@ -0,0 +1,244 @@
+
+tail -n +3 primary.catalog2.invalid.zone > catalog2.invalid.zone.pri
+tail -n +3 secondary2.catalog2.invalid.zone > catalog2.invalid.zone.sec2
+if ! diff catalog1.invalid.zone.pri catalog1.invalid.zone.sec ; then
```
If this is to check the just created files, for catalog2, here is an edit. It has the filenames copied from catalog1.
```suggestion
if ! diff catalog2.invalid.zone.pri catalog2.invalid.zone.sec2 ; then
```

wcawijngaards commented on configparser.y:
This file has reduce/reduce conflicts, that is output from bison in the ci integration run.

wtoorop replied:
Resolved in commit 7f6377fb

wcawijngaards commented on configparser.y:
```diff
@@ -1035,7 +1048,30 @@ pattern_or_zone_option:
+      } else {
+        region_recycle(cfg_parser->opt->region, dname, dname_total_size(dname));
+      }
+    };
```
Possibly moving this code block for `VAR_CATALOG_PRODUCER_ZONE` to the list of code blocks at `pattern_option:` could remove the conflicts that the parser outputs.

wcawijngaards commented on xfrd-catalog-zones.c:
```diff
@@ -0,0 +1,1275 @@
+	 * i.e. label_plus_dname(label1, label_plus_dname(label2, dname))
+	 */
+	memmove(name.bytes + dname->label_count + 2 + ll,
+		((void*)dname) + 2 + dname->label_count, dname->name_size + 1);
```
This seems to suggest the total size of a dname_type is 2+label_count+name_size+1, but I thought the size of dname_type is 2+label_count+name_size. Without the 1 extra in the length copied? That does not exist in a dname_type, because that ends after the uint8_t name[name_size]; I thought?
```suggestion
	memmove(name.bytes + dname->label_count + 2 + ll,
		((void*)dname) + 2 + dname->label_count, dname->name_size);
```

wtoorop replied:
I think you're right indeed. It probably didn't matter that much because there is a certain amount of overhead in the static buffer.

wcawijngaards replied:
But not in the source struct, there may not be overhead there.

wcawijngaards commented on xfrd-catalog-zones.c:
```diff
@@ -0,0 +1,1275 @@
+
+			    /* single rdata atom with single TXT rdata field */
+			||  ((uint8_t*)(rrset->rrs[i].rdatas[0].data + 1))[0]
+			  != (uint8_t) (rrset->rrs[i].rdatas[0].data[0]-1))
```
The comparison should perhaps not be downcasted to uint8_t but instead upcast to uint16_t. The rdata length could, for malformed records, be very long, and only comparing the lower 8 bits may then give the wrong outcome, perhaps?
```suggestion
			||  (uint16_t)(((uint8_t*)(rrset->rrs[i].rdatas[0].data + 1))[0])
			  != (uint16_t) (rrset->rrs[i].rdatas[0].data[0]-1))
```

wcawijngaards commented on xfrd-catalog-zones.c:
```diff
@@ -0,0 +1,1275 @@
+			   */
+		else {
+			int cmp = 0;
+#ifndef NDEBUF
```
Did you mean `NDEBUG`?
```suggestion
#ifndef NDEBUG
```

wcawijngaards commented on xfrd-catalog-zones.c:
```diff
@@ -0,0 +1,1275 @@
+				 */
+				struct catalog_member_zone* to_delete =
+					cursor_cmz(cursor);
+#ifndef NDEBUF
```
Did you mean `NDEBUG`?
```suggestion
#ifndef NDEBUG
```

wcawijngaards commented on xfrd-catalog-zones.c:
```diff
@@ -0,0 +1,1275 @@
+						dname_to_string(
+							cursor_member_id(cursor),
+							NULL)));
+			};
```
The `;` semicolon is not needed here, but also harmless.
```suggestion
			}
```

wcawijngaards reviewed (approved):
This change looks okay to me, there are some issues in the list of suggestions. The consumer and producer parts look nice to have for catalog zones.

wtoorop commented on nsd.conf.5.in:
```diff
@@ -987,10 +987,52 @@ Feed updated zone to verifier over standard input. Default is
+one catalog consumer zone is configured, none of them will be processed.
+Member zones of the catalog will use the pattern specified by the group
+property, or if a group property is missing or invalid, the pattern specified
+by the \fBcatalog-member-pattern\fR option is used. Group properties are valid
```
```suggestion
by the \fBcatalog\-member\-pattern\fR option is used. Group properties are valid
```
Escape dashes in the man page (because it's also escaped elsewhere in the manpage)

wtoorop committed:
Apply suggestions from code review

Thanks Wouter! I'll resolve the remaining reduce/reduce conflicts from command line.

Co-authored-by: Wouter Wijngaards <wcawijngaards@users.noreply.github.com>

wtoorop committed:
Resolve reduce/reduce conflict in configparser

wtoorop committed:
Recommend to use query-acl with catalog zones

wtoorop committed:
ifdef logging the role of processes

k0ekk0ek commented on dname.c:
```diff
@@ -392,6 +392,12 @@ const char *
+}
+
+const char *
+dname_to_string_buf(const dname_type *dname, const dname_type *origin, char *buf)
```
Please update the signature to either incorporate the required length, i.e. `MAXDOMAINLEN * 5` `char buf[MAXDOMAINLEN * 5]` or update the implementation to handle a dynamic length and change the signature to include it.

wtoorop replied:
Done in commit 5a4b474a

k0ekk0ek reviewed (changes requested):
Still looking into this, but thought I'd post one initial remark.

k0ekk0ek commented on util.c:
```diff
@@ -63,6 +63,16 @@ static log_function_type *current_log_function = log_file;
 int log_time_asc = 1;
 
+#ifdef USE_LOG_PROCESS_ROLE
+static const char *global_process_role = NULL;
```
Please don't introduce more static global variables.

wtoorop replied:
Done in commit 5a4b474a

k0ekk0ek commented on difffile.h:
```diff
@@ -66,6 +66,14 @@ int add_RR(namedb_type* db, const dname_type* dname,
+/* apply the xfr file identified by xfrfilenr to zone */
+int apply_ixfr_for_zone(struct nsd* nsd, zone_type* zone, FILE* in,
+        struct nsd_options* opt, udb_base* taskudb, udb_ptr* last_task,
+        uint32_t xfrfilenr, uint8_t called_from);
```
I don't really like this, especially the `called_from` argument. Before xfrd did not have any knowlegde about `zone_type` I believe. I expect this to pull in a bit much for what we need it to do(?)

wtoorop replied:
I agree this is a bit ugly yes. The reason for `called_from` is to prevent exit() when called from xfrd. Calling exit() is just fine and the way to fail a reload when called from main. I think I can change this by letting the function always return, and doing the exit() based on return code from main. Okay with that strategy @k0ekk0ek ?

wtoorop replied:
Resolved in 8b93cc99

k0ekk0ek commented on options.h:
```diff
@@ -308,6 +312,10 @@ struct pattern_options {
 	uint8_t verifier_feed_zone_is_default;
 	int32_t verifier_timeout;
 	uint8_t verifier_timeout_is_default;
+	uint8_t catalog_role;
+	uint8_t catalog_role_is_default;
+	const char* catalog_member_pattern;
+	const char* catalog_producer_zone;
```
Back when the first PR got made, we kind-of decided catalog properties must be set on a zone. A reason for that being so that you cannot all of a sudden break things simply by re-applying patterns. Why should this be part of a pattern?

wtoorop replied:
I evaluated this extensively. Have a look at the catzones.tdir and the text in the man page to see how this all works. The re-applying patterns issues are also covered in catzones.tdir.
Short summary: A catalog producer (catalog_role == CATALOG_ROLE_PRODUCER) is only functional when in a zone clause which is checked by evaluating the value for "catalog_producer_zone" (which only makes sense in a pattern and not in a zone; i.e. catalog member zones are always instantiated through patterns).
Currently only a single catalog consumer is allowed, but this may change in the future. In that future extension there is no reason why you would not have a catalog member zone which is a catalog consumer itself. But that's the future; we do it step by step. Currently it at least allows the user to configure a consumer zone in all the for nsd conventional manner.

k0ekk0ek commented on xfrd-catalog-zones.c:
```diff
@@ -0,0 +1,1275 @@
+		id_label[hex_ntop( (void *)&new_id
+		                 , sizeof(uint32_t)
+		                 , id_label
+		                 , sizeof(id_label))] = 0;
```
Please don't do this. We expect a fixed constant to be returned. `2 * srclength` to be precise. Please just use that when null-terminating. The implementation may just change and then all of a sudden we introduce complexity that's terribly hard to link to `hex_ntop`.

wtoorop replied:
resolved in 5a4b474a

k0ekk0ek commented on xfrd-catalog-zones.c:
```diff
@@ -0,0 +1,1275 @@
+	 */
+	enum { try_to_add, retry_to_add, just_add } mode;
+
+	if (!consumer_zone)
```
Shouldn't this be an `assert`?

wtoorop replied:
There is:
`xfrd_process_catalog_consumer_zone(xfrd_one_catalog_consumer_zone());` on line 439 above. `xfrd_one_catalog_consumer_zone()` can return NULL. I can make it an if there and turn this into an assert if your want. Do you want that?

wtoorop replied:
Addressed in ac967745

wtoorop commented on xfrd-catalog-zones.c:
```diff
@@ -0,0 +1,1275 @@
+		id_label[hex_ntop( (void *)&new_id
+		                 , sizeof(uint32_t)
+		                 , id_label
+		                 , sizeof(id_label))] = 0;
```
```suggestion
		hex_ntop((void*)&new_id, sizeof(uint32_t), id_label, sizeof(id_label));
		id_label[sizeof(uint32_t)*2] = 0;
```

wtoorop commented on dname.c:
```diff
@@ -392,6 +392,12 @@ const char *
+}
+
+const char *
+dname_to_string_buf(const dname_type *dname, const dname_type *origin, char *buf)
```
```suggestion
dname_to_string_buf(const dname_type *dname, const dname_type *origin, char buf[MAXDOMAINLEN * 5])
```

wtoorop commented on util.c:
```diff
@@ -63,6 +63,16 @@ static log_function_type *current_log_function = log_file;
+log_set_process_role(const char *process_role)
+{
+	global_process_role = process_role;
+}
```
```suggestion
void
log_set_process_role(const char *process_role)
{
	global_ident = process_role;
}
```

wtoorop commented on util.c:
```diff
@@ -160,7 +170,13 @@ log_file(int priority, const char *message)
+#else
+			global_ident,
+#endif
+			(int) getpid(), priority_text, message);
```
```suggestion
			global_ident, (int) getpid(), priority_text, message);
```

wtoorop commented on dname.h:
```diff
@@ -346,6 +346,19 @@ label_next(const uint8_t *label)
+ */
+const char *dname_to_string_buf(const dname_type *dname,
+                                const dname_type *origin,
+                                char *buf);
```
```suggestion
                                char buf[MAXDOMAINLEN * 5]);
```

wtoorop committed:
Apply suggestions from code review

Co-authored-by: Wouter Wijngaards <wcawijngaards@users.noreply.github.com>

wtoorop committed:
Update options.c

Recover line to report an error

Co-authored-by: Wouter Wijngaards <wcawijngaards@users.noreply.github.com>

wtoorop committed:
Fix missing MAXDOMAINLEN symbol ...

... now used in a function prototype in dname.h
+ Some clarifying comments in the label_plus_dname() function

wtoorop committed:
Merge branch 'features/xfrd-catz' of github.com:NLnetLabs/nsd into features/xfrd-catz

wtoorop committed:
xfrd_process_catalog_consumer_zone() argument may not be NULL

wtoorop committed:
Remove passing on process role to apply_ixfr_for_zone function

wtoorop committed:
Merge branch 'master' into features/xfrd-catz

wtoorop committed:
ChangeLog and RELNOTES for the Catalog zones merge
